### ISSUE TITLE
plugin/route53: upstream no longer works when configured as self for recursion

### ISSUE THREAD

**Issue by mattlqx** [coredns/coredns#2652]
I got around to trying out CoreDNS 1.4.0 today. I'm currently running 1.2.6 in production and my config works as I'd expect (well, with some already reported issues). I have a Route 53 zone gathered by the route53 plugin.

Corefile:
```
example.com {
  route53 example.com.:Z2CJ10OEFZZZZZ {
    upstream 127.0.0.1:53
    credentials default /configs/credentials
  }
}

. {
  forward . 127.0.0.1:253
}
```

Unbound is running on :253 for recursion. I have upstream pointed to use the port of coredns itself so that the CNAMEs being resolved will run through the full list of configured zones because a CNAME may point to an internal name (also proxied by coredns) or an external name (like in the case of an Amazon LB, which would be a recursive query).

Resolving a name that is an Amazon LB works on 1.2.6, but no longer works on 1.4.0.

Working query against 1.2.6:
```
# dig test.example.com @127.0.0.1 -p 53

; <<>> DiG 9.11.3-1ubuntu1.5-Ubuntu <<>> test.example.com @127.0.0.1 -p 53
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 44400
;; flags: qr rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 4, ADDITIONAL: 0

;; QUESTION SECTION:
;test.example.com.		IN	A

;; ANSWER SECTION:
test.example.com.	4	IN	CNAME	test-2051344828.us-west-2.elb.amazonaws.com.
test-2051344828.us-west-2.elb.amazonaws.com. 4 IN A 1.2.3.4
test-2051344828.us-west-2.elb.amazonaws.com. 4 IN A 5.6.7.8

;; AUTHORITY SECTION:
example.com.		4	IN	NS	ns-1259.awsdns-29.org.
example.com.		4	IN	NS	ns-199.awsdns-24.com.
example.com.		4	IN	NS	ns-633.awsdns-15.net.
example.com.		4	IN	NS	ns-2014.awsdns-59.co.uk.

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Mon Mar 04 10:16:44 PST 2019
;; MSG SIZE  rcvd: 455
```

Same config, non-working query on 1.4.0:

```
# dig test.example.com @127.0.0.1 -p 353

; <<>> DiG 9.11.3-1ubuntu1.5-Ubuntu <<>> test.example.com @127.0.0.1 -p 353
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 27690
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: 4c5c5aff50311445 (echoed)
;; QUESTION SECTION:
;test.example.com.		IN	A

;; Query time: 0 msec
;; SERVER: 127.0.0.1#353(127.0.0.1)
;; WHEN: Mon Mar 04 10:16:46 PST 2019
;; MSG SIZE  rcvd: 63
```

I'm not seeing anything show up in the container log aside from the initial loading messages. I'm guessing this is either related to EDNS or recent changes to plugins recursion availability.

**Comment by fturib** [comment]
> `# dig test.example.com @127.0.0.1 -p 353`
> ; <<>> DiG 9.11.3-1ubuntu1.5-Ubuntu <<>> test.example.com @127.0.0.1 -p 353

Should you test on port 53 ? (instead of 353?)
Not sure what is running on 353 ... but your corefile is expecting on port 53

See [release notes](https://coredns.io/2019/01/13/coredns-1.3.1-release/) for deprecation of `upstream`. Take care to the fact that, from now, you cannot provide the upstream server, resolution of the CNAME occurs internally on the same server.

**Comment by mattlqx** [comment]
I was running 1.4.0 on the same host on port 353 using the same config file just as an illustration.

As for removal of upstream, I couldn’t get the CNAMEs in the zone to resolve properly in either version without the upstream statement which is why I configured it explicitly. That worked fine in 1.2.x. I’ll have to get pastes of the query results of that but if I recall it was the same kind of SERVFAIL scenario.

**Comment by mattlqx** [comment]
Okay, I've taken some time to create some publicly available scenarios to show this in action.

I have a domain I registered for this (proofofconcept.me). I created a public and private zone for it on Route 53 and spun up three CoreDNS docker containers. All the containers will share an Unbound container (port 253) for recursion.

Spun up via:
```
docker run -it -d --name secns.unbound -e INTERFACE=0.0.0.0@253 -e DO_IPV6=no -e VERBOSITY=1 --net host secns/unbound:1.8.1
docker run -it -d --name 'coredns-1.2.6' -v /home/mattlqx/coredns:/configs --net host coredns/coredns:1.2.6 -conf /configs/Corefile
docker run -it -d --name 'coredns-1.2.6-noup' -v /home/mattlqx/coredns:/configs --net host coredns/coredns:1.2.6 -conf /configs/Corefile-1.4 -dns.port 353
docker run -it -d --name 'coredns-1.4.0' -v /home/mattlqx/coredns:/configs --net host coredns/coredns:1.4.0 -conf /configs/Corefile-1.4 -dns.port 354
```

Within the zones on Route 53, I have couple records:
- an overlapping TXT and A record `simple.proofofconcept.me`
- a CNAME to an AWS ELB `elb-public.proofofconcept.me` and `elb-private.proofofconcept.me` in their respective zones
- a root A record (`proofofconcept.me`) pointing to the AWS EC2 instance running the Docker containers

Corefiles are setup to use records from the private zone first and then fallback to the public zone.

#### Scenario 1 - CoreDNS 1.2.6 (port 53) using self as upstream:

Corefile
```
proofofconcept.me {
  route53 proofofconcept.me.:Z3M5VAIPSRDR8Z proofofconcept.me.:Z1OMBJKY7Y7LLQ {
    upstream 127.0.0.1:53
    credentials default /configs/credentials
  }
}

. {
  forward . 127.0.0.1:253
}
```

Result - Good
```
$ dig elb-public.proofofconcept.me @proofofconcept.me

; <<>> DiG 9.10.6 <<>> elb-public.proofofconcept.me @proofofconcept.me
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 11264
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 4, ADDITIONAL: 0

;; QUESTION SECTION:
;elb-public.proofofconcept.me.	IN	A

;; ANSWER SECTION:
elb-public.proofofconcept.me. 300 IN	CNAME	poc-test-1842215533.us-west-2.elb.amazonaws.com.
poc-test-1842215533.us-west-2.elb.amazonaws.com. 21 IN A 52.11.117.219
poc-test-1842215533.us-west-2.elb.amazonaws.com. 21 IN A 35.161.230.29

;; AUTHORITY SECTION:
proofofconcept.me.	172800	IN	NS	ns-920.awsdns-51.net.
proofofconcept.me.	172800	IN	NS	ns-1132.awsdns-13.org.
proofofconcept.me.	172800	IN	NS	ns-1920.awsdns-48.co.uk.
proofofconcept.me.	172800	IN	NS	ns-421.awsdns-52.com.

;; Query time: 42 msec
;; SERVER: 34.222.212.8#53(34.222.212.8)
;; WHEN: Thu Mar 07 09:37:13 MST 2019
;; MSG SIZE  rcvd: 469
```

#### Scenario 2 - CoreDNS 1.2.6 (port 353) without upstream statement:

Corefile
```
proofofconcept.me {
  route53 proofofconcept.me.:Z3M5VAIPSRDR8Z proofofconcept.me.:Z1OMBJKY7Y7LLQ {
    credentials default /configs/credentials
  }
}

. {
  forward . 127.0.0.1:253
}
```

Result - Bad
```
$ dig elb-public.proofofconcept.me @proofofconcept.me -p 353

; <<>> DiG 9.10.6 <<>> elb-public.proofofconcept.me @proofofconcept.me -p 353
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 32147
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;elb-public.proofofconcept.me.	IN	A

;; Query time: 40 msec
;; SERVER: 34.222.212.8#353(34.222.212.8)
;; WHEN: Thu Mar 07 09:43:16 MST 2019
;; MSG SIZE  rcvd: 57
```

#### Scenario 3 - CoreDNS 1.4.0 (port 354) without upstream statement (same as #2):

Corefile
```
proofofconcept.me {
  route53 proofofconcept.me.:Z3M5VAIPSRDR8Z proofofconcept.me.:Z1OMBJKY7Y7LLQ {
    credentials default /configs/credentials
  }
}

. {
  forward . 127.0.0.1:253
}
```

Result - Bad
```
$ dig elb-public.proofofconcept.me @proofofconcept.me -p 354

; <<>> DiG 9.10.6 <<>> elb-public.proofofconcept.me @proofofconcept.me -p 354
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 32162
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;elb-public.proofofconcept.me.	IN	A

;; Query time: 41 msec
;; SERVER: 34.222.212.8#354(34.222.212.8)
;; WHEN: Thu Mar 07 09:44:38 MST 2019
;; MSG SIZE  rcvd: 57
```

To sum up, without using `upstream 127.0.0.1:53`, I can't get CNAME resolution outside of the Route 53 zones working. I've tried a few other things that included a fallthrough and a forward inside the `proofofconcept.me` server block after the route53 plugin and still no dice.

While setting this up and getting the pastes for this, I'm also noticing that CoreDNS stops servicing queries at all shortly after these SERVFAIL responses are served. The daemons are still running, no log messages are made after first loading up, but it fails to respond at all to any queries approximately 30 seconds after the SERVFAIL response.

**Comment by dilyevsky** [comment]
I'm pretty sure this is because of #2436. You still need `upstream` stanza but no args are supported any longer so should be:

```
proofofconcept.me {
  route53 proofofconcept.me.:Z3M5VAIPSRDR8Z proofofconcept.me.:Z1OMBJKY7Y7LLQ {
    credentials default /configs/credentials
    upstream
  }
}

. {
  forward . 127.0.0.1:253
}
```

**Comment by mattlqx** [comment]
Same result with the bare `upstream` unfortunately.

**Comment by dilyevsky** [comment]
Hm, ok let me see if i can reproduce this in my env. Thanks for report!

**Comment by chrisohaver** [comment]
> You still need upstream stanza but no args are supported any longer so should be:

I believe the behavior in 1.4.0 is to completely ignore the `upstream` option and always upstream to self even if `upstream` is not present in the config.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] [plugin/route..." ]
>> You still need upstream stanza but no args are supported any longer so should be:
>
>I believe the behavior in 1.4.0 is to completely ignore the `upstream` option and always upstream to self even if `upstream` is not present in the config.

I'm not sure if we made that switch *yet*. However the plan is to do this; i.e. 
we ask the coredns process itself, regardless of `upstream` being present in the 
Corefile. If you have a forward/proxy/resolver that allows this to be resolved, 
it will be.

(in the route53 code you still need a *upstream somewhere to make this work)

**Comment by dilyevsky** [comment]
> (in the route53 code you still need a *upstream somewhere to make this work)

Yes, this is correct. I will send a PR to address this.

**Comment by dilyevsky** [comment]
@mattlqx actually i tried this today and this stanza is causing loop in your query path:
```
. {
  forward . 127.0.0.1:253
}
```

You can verify this yourself using `loop` option.

It should be pointing to an upstream resolver like so:

```
. {
  forward . 8.8.8.8
}
```

**Comment by mattlqx** [comment]
Port 253 is the unbound container though, not coredns itself, as described at the top of https://github.com/coredns/coredns/issues/2652#issuecomment-470612340.

**Comment by dilyevsky** [comment]
Ah I see. Are you sure that upstream works correctly? I'm able to resolve CNAMEs with the following config with Coredns I built from the latest master:

```
my.zone {
  route53 my.zone:ZID {
    credentials my-zone aws-creds
    upstream
  }
  errors
}

. {
  loop
  forward . 8.8.8.8
}
```

**Comment by mattlqx** [comment]
Yep. Here's unedited output. The simplest case is elb-public.proofofconcept.me. Unbound sees it no problem. And to prove that an A record works fine, there's also a lookup of the root record.

```
root@proofofconcept:~# cat coredns/Corefile-1.4
proofofconcept.me {
  route53 proofofconcept.me.:Z3M5VAIPSRDR8Z proofofconcept.me.:Z1OMBJKY7Y7LLQ {
    credentials default /configs/credentials
    upstream
  }
  errors
}

. {
  loop
  forward . 127.0.0.1:253
}
root@proofofconcept:~# docker logs coredns-1.4.0
.:354
proofofconcept.me.:354
2019-03-29T21:36:40.013Z [INFO] CoreDNS-1.4.0
2019-03-29T21:36:40.013Z [INFO] linux/amd64, go1.12, 8dcc7fc
CoreDNS-1.4.0
linux/amd64, go1.12, 8dcc7fc
root@proofofconcept:~# dig proofofconcept.me @proofofconcept.me -p 354

; <<>> DiG 9.11.3-1ubuntu1.5-Ubuntu <<>> proofofconcept.me @proofofconcept.me -p 354
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 22188
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 4, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: 6f6e2513004a1fe2 (echoed)
;; QUESTION SECTION:
;proofofconcept.me.		IN	A

;; ANSWER SECTION:
proofofconcept.me.	300	IN	A	34.222.212.8

;; AUTHORITY SECTION:
proofofconcept.me.	172800	IN	NS	ns-0.awsdns-00.com.
proofofconcept.me.	172800	IN	NS	ns-1024.awsdns-00.org.
proofofconcept.me.	172800	IN	NS	ns-1536.awsdns-00.co.uk.
proofofconcept.me.	172800	IN	NS	ns-512.awsdns-00.net.

;; Query time: 0 msec
;; SERVER: 34.222.212.8#354(34.222.212.8)
;; WHEN: Fri Mar 29 21:43:58 UTC 2019
;; MSG SIZE  rcvd: 297
root@proofofconcept:~# dig elb-public.proofofconcept.me @proofofconcept.me -p 253

; <<>> DiG 9.11.3-1ubuntu1.5-Ubuntu <<>> elb-public.proofofconcept.me @proofofconcept.me -p 253
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 53495
;; flags: qr rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;elb-public.proofofconcept.me.	IN	A

;; ANSWER SECTION:
elb-public.proofofconcept.me. 195 IN	CNAME	poc-test-1842215533.us-west-2.elb.amazonaws.com.
poc-test-1842215533.us-west-2.elb.amazonaws.com. 60 IN A 52.11.117.219
poc-test-1842215533.us-west-2.elb.amazonaws.com. 60 IN A 35.161.230.29

;; Query time: 25 msec
;; SERVER: 34.222.212.8#253(34.222.212.8)
;; WHEN: Fri Mar 29 21:39:28 UTC 2019
;; MSG SIZE  rcvd: 150

root@proofofconcept:~# dig elb-public.proofofconcept.me @proofofconcept.me -p 354

; <<>> DiG 9.11.3-1ubuntu1.5-Ubuntu <<>> elb-public.proofofconcept.me @proofofconcept.me -p 354
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 34156
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: 0ef4be973671a3ce (echoed)
;; QUESTION SECTION:
;elb-public.proofofconcept.me.	IN	A

;; Query time: 0 msec
;; SERVER: 34.222.212.8#354(34.222.212.8)
;; WHEN: Fri Mar 29 21:39:32 UTC 2019
;; MSG SIZE  rcvd: 69

```

About 30 seconds after that SERVFAIL, coredns no longer responds to requests. Is there an easy way for me to do debug what's going on there when it goes unresponsive?

**Comment by pryorda** [comment]
Any updates on any of this?

**Comment by mattlqx** [comment]
I've upgraded part of my environment to 1.6.0 and this doesn't appear to be an issue any longer. I'm going to give it 24 hours and call it fixed if all looks good still.

**Comment by dilyevsky** [comment]
That sort of confirms my suspicion that this is either an upstream dependency issue or some weird side-effect bug. @mattlqx which version were you using previously? will it be too much work to bisect to commit that fixes this?

**Comment by ricardbejarano** [comment]
I'm having a similar problem when using the `hosts` plugin:

### Without `hosts`

#### Corefile
```
.:1053 {
	log
	errors
	forward . 1.1.1.1
}
```

#### Query
```
$ dig -p 1053 @127.0.0.1 bejarano.io

; <<>> DiG 9.10.6 <<>> -p 1053 @127.0.0.1 bejarano.io
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 52201
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;bejarano.io.			IN	A

;; ANSWER SECTION:
bejarano.io.		172	IN	A	104.28.3.167
bejarano.io.		172	IN	A	104.28.2.167

;; Query time: 15 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Wed Jul 31 00:18:36 CEST 2019
;; MSG SIZE  rcvd: 94
```

#### CoreDNS output
```
$ ./coredns
.:1053
2019-07-30T22:21:55.621Z [INFO] CoreDNS-1.6.0
2019-07-30T22:21:55.625Z [INFO] linux/amd64, go1.12.7, 
CoreDNS-1.6.0
linux/amd64, go1.12.7, 
2019-07-30T22:21:57.436Z [INFO] 127.0.0.1:60419 - 8772 "A IN bejarano.io. udp 40 false 4096" NOERROR qr,rd,ra,ad 94 0.0201465s```
```

### With `hosts`

#### Corefile
```
.:1053 {
	log
	errors
	hosts /etc/hosts.local
	forward . 1.1.1.1
}
```

#### Query
```
$ dig -p 1053 @127.0.0.1 bejarano.io

; <<>> DiG 9.10.6 <<>> -p 1053 @127.0.0.1 bejarano.io
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 22683
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;bejarano.io.			IN	A

;; Query time: 1 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Wed Jul 31 00:25:35 CEST 2019
;; MSG SIZE  rcvd: 40
```

#### CoreDNS output
```
$ ./coredns
.:1053
2019-07-30T22:25:32.524Z [INFO] CoreDNS-1.6.0
2019-07-30T22:25:32.526Z [INFO] linux/amd64, go1.12.7, 
CoreDNS-1.6.0
linux/amd64, go1.12.7,
```

It may be relevant to note that the `log` module **does not** log the incoming request when it `SERVFAIL`s.

**Comment by mattlqx** [comment]
@dilyevsky it wasn't too bad.

Results:
tag v1.2.6 base
working -> broken   9c16ed1d14f570f27b4ff9cb89845a3a1548a776
broken -> working   53f3f0b666821588e721ceeea4766b76333b668b

Looks like v1.5.0 was the first one with the fix. I have been running v1.2.6 for a while. I tried to upgrade to v1.4.0 but that's where I encountered the problem that lead to this issue.

**Comment by dilyevsky** [comment]
That's awesome, thanks! I think I see it now - whenever it tried to go to upstream, it would try to use empty context. 53f3f0b666821588e721ceeea4766b76333b668b fixed this by passing in context from caller.

Lets keep this open for now - I'd like to add regression test that hits the upstream path.

**Comment by miekg** [comment]
The host plugins panics, but the code recovers. This is way you get
servfail, but no logs.

See recent prs for a fix.
